src/Auth0.OidcClient.Core/Tokens/AsymmetricSignatureVerifier.cs

```diff
@@ -0,0 +1,75 @@
+﻿using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+
+namespace Auth0.OidcClient.Tokens
+{
+    internal class AsymmetricSignatureVerifier : ISignatureVerifier
+    {
+        private readonly IList<JsonWebKey> keys;
+
+        public AsymmetricSignatureVerifier(IList<JsonWebKey> keys)
+        {
+            this.keys = keys;
+        }
+
+        public JwtSecurityToken VerifySignature(string token)
+        {
+            var securityTokenHandler = new JwtSecurityTokenHandler();
+
+            JwtSecurityToken decoded;
+            try
+            {
+                decoded = securityTokenHandler.ReadJwtToken(token);
+            }
+            catch (ArgumentException e)
+            {
+                throw new IdTokenValidationException("ID token could not be decoded.", e);
+            }
+
+            if (decoded.Header.Alg != "RS256")
+                throw new IdTokenValidationException($"Signature algorithm of \"{decoded.Header.Alg }\" is not supported. Expected the ID token to be signed with \"RS256\".");
+
+            var publicKey = FindPublicKeyByKid(decoded.Header.Kid);
+
+            return (JwtSecurityToken) ValidateTokenSignatureWithKey(token, securityTokenHandler, publicKey);
+        }
+
+        internal static SecurityToken ValidateTokenSignatureWithKey(string token, JwtSecurityTokenHandler securityTokenHandler, JsonWebKey key)
+        {
+            try
+            {
+                var validationParameters = new TokenValidationParameters
+                {
+                    ValidateAudience = false,
+                    ValidateIssuer = false,
+                    ValidateActor = false,
+                    ValidateLifetime = false,
+                    ValidateTokenReplay = false,
+
+                    RequireSignedTokens = true,
+                    ValidateIssuerSigningKey = true,
+                    IssuerSigningKey = key,
+                };
+
+                securityTokenHandler.ValidateToken(token, validationParameters, out var verifiedToken);
+                return verifiedToken;
+            }
+            catch (SecurityTokenException e)
+            {
+                throw new IdTokenValidationException("Invalid ID token signature.", e);
+            }
+        }
+
+        private JsonWebKey FindPublicKeyByKid(string kid)
+        {
+            var key = keys.FirstOrDefault(k => k.Kid == kid);
+            if (key == null)
+                throw new IdTokenValidationException($"Could not find a public key for kid \"{kid}\".");
```

Comment:
**Nit:** This error message is not in the sample. Should this instead throw "Invalid ID token signature"?

Reply:
I guess if a Pub key is not found it won't be possible to verify the signature and in the end, the token would be rejected. I'm going to be on the same situation with Android. Do we want to add this message to the error messages table?

Reply:
Started a topic internally to discuss this message.